File: code/java/2817.java
// https://leetcode.com/problems/minimum-absolute-difference-between-elements-with-constraint/
class Solution {
    public int minAbsoluteDifference(List<Integer> nums, int x) {
        int res = Integer.MAX_VALUE;
        TreeSet<Integer> set = new TreeSet<>();
        for (int i = x; i < nums.size(); ++i) {
        set.add(nums.get(i - x));
        Integer high = set.ceiling(nums.get(i)), low = set.floor(nums.get(i));
        if (high != null)
            res = Math.min(res, Math.abs(nums.get(i) - high));
        if (low != null)
            res = Math.min(res, Math.abs(nums.get(i) - low));
        }
        return res;
    }
}

File: code/python/2817.py
# https://leetcode.com/problems/minimum-absolute-difference-between-elements-with-constraint/
class Solution:
    def minAbsoluteDifference(self, nums: List[int], x: int) -> int:
        sorted_nums = SortedList(nums[x:])
        n = len(nums)
        res = float('inf')
        for i in range(0, n-x):
            num = nums[i]
            idx = sorted_nums.bisect_left(num)
            if idx < len(sorted_nums):
                res = min(res, abs(sorted_nums[idx] - num))
            if idx > 0:
                res = min(res, abs(sorted_nums[idx - 1] - num))
            if res == 0:
                return 0
            sorted_nums.remove(nums[i+x])
        return res
